include/pybind11/pytypes.h

```diff
@@ -375,31 +388,101 @@ PYBIND11_NAMESPACE_END(detail)
+                    throw std::runtime_error(
+                        "FATAL failure building pybind11::detail::error_already_set what()");
+                }
+            } catch (...) {
+                // Terminating the process, to not mask the original error by errors in the error
+                // handling. Reporting the original error on stderr & stdout. Intentionally using
+                // the Python C API directly, to maximize reliability.
+                std::string msg
+                    = "FATAL failure building pybind11::detail::error_already_set what(): ";
+                if (m_type.ptr() == nullptr) {
+                    msg += "PYTHON_EXCEPTION_TYPE_IS_NULLPTR";
+                } else {
+                    const char *class_name = detail::obj_class_name(m_type.ptr());
+                    if (class_name == nullptr) {
+                        msg += "PYTHON_EXCEPTION_CLASS_NAME_IS_NULLPTR";
+                    } else {
+                        msg += class_name;
+                    }
+                }
+                msg += ": ";
+                PyObject *val_str = PyObject_Str(m_value.ptr());
+                if (val_str == nullptr) {
+                    msg += "PYTHON_EXCEPTION_VALUE_IS_NULLPTR";
+                } else {
+                    Py_ssize_t utf8_str_size = 0;
+                    const char *utf8_str = PyUnicode_AsUTF8AndSize(val_str, &utf8_str_size);
+                    if (utf8_str == nullptr) {
+                        msg += "PYTHON_EXCEPTION_VALUE_AS_UTF8_FAILURE";
+                    } else {
+                        msg += '"' + std::string(utf8_str, static_cast<std::size_t>(utf8_str_size))
+                               + '"';
+                    }
+                }
+                // Intentionally using C calls to maximize reliability
+                // (and to avoid #include <iostream>).
+                fprintf(stderr, "%s [STDERR]\n", msg.c_str());
+                fflush(stderr);
+                fprintf(stdout, "%s [STDOUT]\n", msg.c_str());
+                fflush(stdout);
+                std::terminate();
+            }
+        }
+        return m_lazy_what.c_str();
+    }
+
+    /// Restores the currently-held Python error (which will clear the Python error indicator first
+    /// if already set).
+    /// NOTE: This function will not necessarily restore the original Python exception, but may
+    ///       restore the normalized exception if what() or discard_as_unraisable() were called
+    ///       prior to restore().
     void restore() {
-        PyErr_Restore(m_type.release().ptr(), m_value.release().ptr(), m_trace.release().ptr());
+        // As long as this type is copyable, there is no point in releasing m_type, m_value,
+        // m_trace, but simply holding on the the references makes it possible to produce
+        // what() even after restore().
+        PyErr_Restore(m_type.inc_ref().ptr(), m_value.inc_ref().ptr(), m_trace.inc_ref().ptr());
```

Comment:
Hmm, I am wondering if referring to these types after Err_Restore is UB:
`Set the error indicator from the three objects. If the error indicator is already set, it is cleared first. If the objects are NULL, the error indicator is cleared. Do not pass a NULL type and non-NULL value or traceback. The exception type should be a class. Do not pass an invalid exception type or value. (Violating these rules will cause subtle problems later.) This call takes away a reference to each object: you must own a reference to each object before the call and after the call you no longer own these references. (If you don’t understand this, don’t use this function. I warned you.)`